src/mame/mit/cadr_disk.cpp

```diff
@@ -0,0 +1,361 @@
+					m_data_space->write_dword((page * 256) + (i / 4), data);
+				}
+				if (BIT(ccw, 0) && max_ccws)
+				{
+					sector++;
+					m_clp++;
+					max_ccws--;
+				}
+				else
+				{
+					max_ccws = 0;
+				}
+				m_last_memory_address = ccw & 0xffffff00;
+			} while (max_ccws);
+			m_block = sector % info.sectors;
+			sector /= info.sectors;
+			m_head = sector % info.heads;
+			m_cyl = sector / info.heads;
+			m_status |= 1;
+			m_disk_timer->adjust(attotime::from_msec(6));
+		}
+		else
+		{
+			m_status |= 0x8200; // The timeout bit is checked for a unit that is not present?
+			m_status |= 1;
+			m_disk_timer->adjust(attotime::from_msec(6));
+		}
+		break;
+
+	case 0x005: // 0005 - At ease
+		m_status &= ~0x04;
+		m_status |= 1;
+		break;
+
+	case 0x009: // 0011 - Write
+		LOG("Start write\n");
+		if (m_harddisk->exists())
+		{
+			u32 max_ccws = 0x10000;
+			const auto &info = m_harddisk->get_info();
+			u32 sector = (m_cyl * info.heads * info.sectors) + (m_head * info.sectors) + m_block;
+			u8 buffer[1024];
+			do
+			{
+				const u32 ccw = m_data_space->read_dword(m_clp);
+				const u16 page = (ccw >> 8) & 0x7fff;
+				LOG("Start write sector %d\n", sector);
+				LOG("Read sector data from %08x\n", page << 8);
+				for (int i = 0; i < 256; i++)
+				{
+					u32 data = m_data_space->read_dword((page * 256) + i);
+					buffer[(i * 4) + 3] = (data >> 24) & 0xff;
+					buffer[(i * 4) + 2] = (data >> 16) & 0xff;
+					buffer[(i * 4) + 1] = (data >> 8) & 0xff;
+					buffer[i * 4] = data & 0xff;
+				}
+				m_harddisk->write(sector, &buffer[0]);
+				if (BIT(ccw, 0) && max_ccws)
+				{
+					sector++;
+					m_clp++;
+					max_ccws--;
+				}
+				else
+				{
+					max_ccws = 0;
+				}
+				m_last_memory_address = ccw & 0xffffff00;
+			} while (max_ccws);
+			m_block = sector % info.sectors;
+			sector /= info.sectors;
+			m_head = sector % info.heads;
+			m_cyl = sector / info.heads;
+			m_status |= 1;
+			m_disk_timer->adjust(attotime::from_msec(6));
+		}
+		break;
+
+	case 0x105: // 0405 - Fault clear
+		LOG("Start fault clear\n");
+		m_status |= 1;
+		break;
+
+	case 0x205: // 1005 - Recalibrate
+		LOG("Start recalibrate\n");
+		m_status |= 5;
+		break;
+
+	case 0x008: // 0010 - Read and compare
+	case 0x004: // 0004 - Seek
+	default:
+		fatalerror("Unknown disk controller command %03x initiated\n", m_command & 0x3ff);
+		break;
+	}
+}
+
+
+u32 cadr_disk_device::read(offs_t offset)
+{
+	switch (offset)
+	{
+	case 0x04:
+		return status_r();
+	case 0x05:
+		return command_list_r();
+	case 0x06:
+		return disk_address_r();
+	case 0x07:
+		return error_correction_r();
+	}
+	return 0xffffffff;
+}
+
+
+void cadr_disk_device::write(offs_t offset, u32 data)
+{
+	switch (offset)
+	{
+	case 0x04:
+		command_w(data);
+		break;
+	case 0x05:
+		command_list_w(data);
+		break;
+	case 0x06:
+		disk_address_w(data);
+		break;
+	case 0x07:
+		start_w(data);
+		break;
+	}
+}
```

Comment:
Would you be better off with a device address map function rather than these dispatch functions?